packages/content-conformance/src/rules/ensure-valid-link-format.js

```diff
@@ -0,0 +1,207 @@
+  executor: {
+    contentFile(file, context) {
+      file.visit(['link', 'definition'], (node) => {
+        const urlObject = new URL(node.url, TEST_URL_ORIGIN)
+        const { hostname, pathname, search = '', hash = '', origin } = urlObject
+        const isRelativePath = origin === TEST_URL_ORIGIN
+
+        // Error if we detect folder-relative links
+        if (node.url.startsWith('.')) {
+          context.report(
+            `Unexpected folder-relative link found: ${node.url}. Ensure this link is an absolute Developer path.`,
+            file,
+            node
+          )
+          return
+        }
+
+        // Check for .io hostname
+        const dotIoLinkProductSlug = Object.keys(
+          PRODUCT_SLUGS_TO_HOST_NAMES
+        ).find((productSlug) => {
+          return (
+            hostname === PRODUCT_SLUGS_TO_HOST_NAMES[productSlug] ||
+            hostname === `www.${PRODUCT_SLUGS_TO_HOST_NAMES[productSlug]}`
+          )
+        })
+        if (dotIoLinkProductSlug) {
+          const devDotBasePaths =
+            PRODUCT_SLUGS_TO_BASE_PATHS[dotIoLinkProductSlug]
+          const isDevDotBasePath =
+            pathname === '/' ||
+            devDotBasePaths.some((devDotBasePath) => {
+              return pathname.startsWith(`/${devDotBasePath}`)
+            })
+          if (isDevDotBasePath) {
+            context.report(
+              `Unexpected link to documentation on \`${
+                PRODUCT_SLUGS_TO_HOST_NAMES[dotIoLinkProductSlug]
+              }\`: \`${
+                node.url
+              }\`. Try replacing it with \`${`/${dotIoLinkProductSlug}${
+                pathname === '/' ? '' : pathname
+              }`}\``,
+              file,
+              node
+            )
+          }
+        }
+
+        // Check for learn hostname
+        if (hostname === LEARN_HOSTNAME) {
+          context.report(
+            `Unexpected link to \`learn.hashicorp.com\`: \`${node.url}\`. Replace it with a relative path internal to Developer with the format: \`/{productSlug}/tutorials/{collectionSlug}/{tutorialSlug}\`.`,
+            file,
+            node
+          )
+        }
+
+        // Check for developer hostname
+        if (hostname === DEV_DOT_HOSTNAME) {
+          context.report(
+            `Unexpected fully-qualified link to \`developer.hashicorp.com\`: \`${node.url}\`. Replace with a relative path internal to Developer. Possibly: \`${pathname}${search}${hash}\`.`,
+            file,
+            node
+          )
+        }
+
+        // handle product-relative paths
+        if (
+          isRelativePath &&
+          // ignore any paths starting with /{productSlug}
+          !pathname.match(PRODUCT_SLUG_PATH_PATTERN) &&
+          // check paths that start with a known docs base path
+          KNOWN_DOCS_BASE_PATHS.some((basePath) =>
+            pathname.startsWith(basePath)
+          )
```

Comment:
I went with this approach so we did not have to maintain a manifest of valid developer routes here. It feels like a potential maintenance headache to have to remember to update this rule if we do add more root-level Developer paths.

This way we assume that non-docs base paths are "valid" in structure, and we can rely on other forms of validation on these links to ensure they are valid in destination.